src/command_modules/azure-cli-consumption/azure/cli/command_modules/consumption/custom.py

```diff
@@ -62,12 +62,71 @@ def cli_consumption_list_reservations_details(client, reservation_order_id, star
     if include_meter_details:
         expand_properties = 'properties/meterDetails'
     else:
         expand_properties = None
 
-    if billing_period_name:
+    if billing_period_name and top:
+        return client.get_by_billing_period(billing_period_name, expand=expand_properties, top=top)
+    elif billing_period_name and not top:
         return client.get_by_billing_period(billing_period_name, expand=expand_properties)
+    elif not billing_period_name and top:
+        return client.get(expand=expand_properties, top=top)
     return client.get(expand=expand_properties)
+
+
+def cli_consumption_list_marketplace(client, billing_period_name=None, start_date=None, end_date=None, top=None):
+    filter_from = None
+    filter_to = None
+    filter_expression = None
+    if start_date and end_date:
+        filter_from = "properties/usageEnd ge \'{}\'".format(start_date.strftime("%Y-%m-%dT%H:%M:%SZ"))
+        filter_to = "properties/usageEnd le \'{}\'".format(end_date.strftime("%Y-%m-%dT%H:%M:%SZ"))
+        filter_expression = "{} and {}".format(filter_from, filter_to)
+
+    if billing_period_name and top:
+        return list(client.list_by_billing_period(billing_period_name, filter=filter_expression, top=top).advance_page())
+    elif billing_period_name and not top:
+        return list(client.list_by_billing_period(billing_period_name, filter=filter_expression))
+    elif not billing_period_name and top:
+        return list(client.list(filter=filter_expression, top=top).advance_page())
+    return client.list(filter=filter_expression)
+
+
+def cli_consumption_list_budgets(client, resource_group_name=None):
+    if resource_group_name:
+        return list(client.list_by_resource_group_name(resource_group_name))
+    return list(client.list())
+
+
+def cli_consumption_show_budget(client, budget_name, resource_group_name=None):
+    if resource_group_name:
+        return client.get_by_resource_group_name(resource_group_name, budget_name)
+    return client.get(budget_name)
+
+
+def cli_consumption_create_budget(client, budget_name, category, amount, time_grain, start_date, end_date, resource_groups=None, resources=None, meters=None, resource_group_name=None):
+    time_period = client.models.BudgetTimePeriod(start_date, end_date)
+    filters = client.models.Filters(resource_groups=resource_groups, resources=resources, meters=meters)
+    parameters = client.models.Budget(category=category, amount=amount, time_grain=time_grain, time_period=time_period, filters=filters, notifications=None)
+    if resource_group_name:
+        return client.create_or_update(resource_group_name, budget_name, parameters)
+    return client.create_or_update(budget_name, parameters)
+
+
+def cli_consumption_update_budget(client, budget_name, category, amount, time_grain, start_date, end_date, e_tag, resource_groups=None, resources=None, meters=None, notifications=None, resource_group_name=None):
```

Comment:
This is a copy/paste/rename of the create command and does not conform to CLI expectations for update commands. 
https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#command-naming-and-behavior-guidance